### PR TITLE
Allow adding report_prefix for gcp provider.

### PR DESCRIPTION
**Pull request by katherine-black**

### CODE REVIEW

**Comment by katherine-black** [comment]
Here are a couple of quick test.

Too long of a prefix
```
$ http POST 0.0.0.0:8000/api/cost-management/v1/providers/ name="GCP Provider" type=GCP billing_source:='{"data_source": {"bucket": "precious-taters", "report_prefix": "taterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaterstaters"}}' authentication:='{"credentials": {"project_id": "crucial-baton-252413"}}'
HTTP/1.1 400 Bad Request
Allow: GET, POST, HEAD, OPTIONS
Content-Length: 140
Content-Type: application/json
Date: Thu, 10 Oct 2019 15:31:35 GMT
Server: WSGIServer/0.2 CPython/3.6.6
Vary: Accept, Origin
X-Frame-Options: SAMEORIGIN

{
    "errors": [
        {
            "detail": "report_prefix length is 168, while the maximum allowed is 64.",
            "source": "billing_source.report_prefix",
            "status": 400
        }
    ]
}
```

No prefix
```
$ http POST 0.0.0.0:8000/api/cost-management/v1/providers/ name="GCP Provider" type=GCP billing_source:='{"data_source": {"bucket": "precious-taters"}}' authentication:='{"credentials": {"project_id": "crucial-baton-252413"}}'
HTTP/1.1 201 Created
Allow: GET, POST, HEAD, OPTIONS
Cache-Control: max-age=0, no-cache, no-store, must-revalidate
Content-Length: 649
Content-Type: application/json
Date: Thu, 10 Oct 2019 15:31:58 GMT
Expires: Thu, 10 Oct 2019 15:31:58 GMT
Server: WSGIServer/0.2 CPython/3.6.6
Vary: Accept, Origin
X-Frame-Options: SAMEORIGIN

{
    "authentication": {
        "credentials": {
            "project_id": "crucial-baton-252413"
        },
        "provider_resource_name": null,
        "uuid": "b58bd0c5-e9df-40ac-8635-0f8ffc474526"
    },
    "billing_source": {
        "bucket": "precious-taters",
        "data_source": {
            "bucket": "precious-taters"
        },
        "uuid": "d85ab86d-97b5-4e33-aa12-dc16b149198d"
    },
    "created_by": {
        "email": "user_dev@foo.com",
        "username": "user_dev",
        "uuid": "64f52104-da49-46fd-ba5c-8e2ad4f235a6"
    },
    "created_timestamp": "2019-10-10T15:31:58.077801Z",
    "customer": {
        "account_id": "10001",
        "date_created": "2019-10-10T14:23:19.937117Z",
        "uuid": "31234dfc-405e-4397-a56e-32733d58e2e9"
    },
    "name": "GCP Provider",
    "type": "GCP",
    "uuid": "9b67350f-9238-4655-9b3a-e5151d006fad"
}
```

Appropriate prefix
```
$ http POST 0.0.0.0:8000/api/cost-management/v1/providers/ name="GCP Provider" type=GCP billing_source:='{"data_source": {"bucket": "precious-taters", "report_prefix": "taters"}}' authentication:='{"credentials": {"project_id": "crucial-baton-252413"}}'
HTTP/1.1 201 Created
Allow: GET, POST, HEAD, OPTIONS
Cache-Control: max-age=0, no-cache, no-store, must-revalidate
Content-Length: 674
Content-Type: application/json
Date: Thu, 10 Oct 2019 15:35:59 GMT
Expires: Thu, 10 Oct 2019 15:35:59 GMT
Server: WSGIServer/0.2 CPython/3.6.6
Vary: Accept, Origin
X-Frame-Options: SAMEORIGIN

{
    "authentication": {
        "credentials": {
            "project_id": "crucial-baton-252413"
        },
        "provider_resource_name": null,
        "uuid": "d73242d9-722d-445f-8aee-be6d2f7c5040"
    },
    "billing_source": {
        "bucket": "precious-taters",
        "data_source": {
            "bucket": "precious-taters",
            "report_prefix": "taters"
        },
        "uuid": "4d3d0653-7e17-4a34-bb56-79b1ea5c9167"
    },
    "created_by": {
        "email": "user_dev@foo.com",
        "username": "user_dev",
        "uuid": "ae0b8150-1011-460e-b895-1759b07ef007"
    },
    "created_timestamp": "2019-10-10T15:35:59.963535Z",
    "customer": {
        "account_id": "10001",
        "date_created": "2019-10-10T15:35:48.943751Z",
        "uuid": "878bf998-2d11-4f5a-b61b-0212e6e83bba"
    },
    "name": "GCP Provider",
    "type": "GCP",
    "uuid": "a1758525-eafc-4c35-a08e-88f858904f1b"
}
```

File: koku/providers/gcp/provider.py [hunk 1]
```diff
@@ -40,6 +40,13 @@ def cost_usage_source_is_reachable(self, credential_name, data_source):
             key = 'billing_source.bucket'
             raise serializers.ValidationError(error_obj(key, e.message))
 
+        report_prefix = data_source.get('report_prefix', '')
```
Comment (werwty) [inline koku/providers/gcp/provider.py]: https://github.com/project-koku/koku/blob/master/koku/api/provider/serializers.py#L159 might be a better place to put this check. I think `cost_usage_source_is_reachable` should only check whether or not GCP bucket is reachable.

File: koku/providers/gcp/provider.py [hunk 2]
```diff
@@ -40,6 +40,13 @@ def cost_usage_source_is_reachable(self, credential_name, data_source):
...
+        if report_prefix:
+            if len(report_prefix) > 64:
+                key = 'billing_source.report_prefix'
+                message = f'report_prefix length is {len(report_prefix)}, while the maximum allowed is 64.'
```
Comment (infinitewarp) [inline koku/providers/gcp/provider.py]: We could make this _look_ more like a standard DRF `CharField` validation error message (even though it's all custom validation) if we use the string:

```
'Ensure this field has no more than {max_length} characters.'
```

File: koku/providers/gcp/provider.py [hunk 3]
```diff
@@ -40,6 +40,13 @@ def cost_usage_source_is_reachable(self, credential_name, data_source):
             key = 'billing_source.bucket'
             raise serializers.ValidationError(error_obj(key, e.message))
 
+        report_prefix = data_source.get('report_prefix', '')
+        if report_prefix:
+            if len(report_prefix) > 64:
```
Comment (infinitewarp) [inline koku/providers/gcp/provider.py]: For easier future revision, let's put this `64` in a module-level or class-level variable and also use it in `test_bucket_report_prefix_too_long` for its too-long value.

File: koku/masu/external/downloader/gcp/gcp_report_downloader.py [hunk 4]
```diff
@@ -206,9 +207,14 @@ def _get_relevant_file_names(self, start_date, end_date):
         dates_as_strings = [date.strftime('%Y-%m-%d') for date in dates]
         available_file_names = self._get_bucket_file_names()
         relevant_file_names = set()
+        prefix = ''
```
Comment (infinitewarp) [inline koku/masu/external/downloader/gcp/gcp_report_downloader.py]: Collapse these lines into a ternary?
```suggestion
        prefix = f'{self.report_prefix}-' if self.report_prefix else ''
```
Just a style suggestion.

**Pull request merged**